File: apps/android-customer/feature/catalog/src/main/java/co/zw/nissangtr/customer/catalog/CatalogBrowseFilters.kt
package co.zw.nissangtr.customer.catalog

import co.zw.nissangtr.customer.rpc.CatalogListItem
import co.zw.nissangtr.ui.shop.ShopFilterState
import co.zw.nissangtr.ui.shop.ShopSortOption

/** Client-side filter + sort on browse rows (web PLP / `applyCatalogFiltersAndSort` parity). */
internal fun applyCatalogFilterSort(
    items: List<CatalogListItem>,
    filter: ShopFilterState,
    sort: ShopSortOption,
    /** When true (default for /shop), keep only qty > 0 and priced > 0. */
    shopStockOnly: Boolean = true,
): List<CatalogListItem> {
    var out = items.asSequence()
    if (shopStockOnly) {
        out = out.filter { item ->
            val qty = item.qty
            val inStock = when {
                qty != null -> qty > 0
                else -> item.stock != co.zw.nissangtr.customer.rpc.StockState.BACKORDER
            }
            inStock && item.usd != null && item.usd > 0
        }
    }
    if (filter.minPrice > 0f || filter.maxPrice < 500f) {
        out = out.filter { item ->
            val usd = item.usd ?: return@filter false
            usd >= filter.minPrice && usd <= filter.maxPrice
        }
    }
    filter.category?.trim()?.takeIf { it.isNotEmpty() }?.let { cat ->
        out = out.filter { item ->
            item.category?.equals(cat, ignoreCase = true) == true
        }
    }
    val list = out.toList()
    return when (sort) {
        ShopSortOption.Relevance -> list
        ShopSortOption.PriceAsc -> list.sortedBy { it.usd ?: Double.MAX_VALUE }
        ShopSortOption.PriceDesc -> list.sortedByDescending { it.usd ?: Double.MIN_VALUE }
        ShopSortOption.NameAsc -> list.sortedBy { it.name.lowercase() }
    }
}

File: apps/ios/GTRCustomer/Features/CatalogBrowseFilters.swift
import Foundation

/// Client-side filter + sort on browse rows (Android / web PLP parity).
struct ShopFilterState: Equatable {
    var minPrice: Float = 0
    var maxPrice: Float = 500
    var category: String?
}

enum ShopSortOption: String, CaseIterable, Identifiable {
    case relevance, priceAsc, priceDesc, nameAsc

    var id: String { rawValue }

    var label: String {
        switch self {
        case .relevance: return "Relevance"
        case .priceAsc: return "Price: low to high"
        case .priceDesc: return "Price: high to low"
        case .nameAsc: return "Name: A–Z"
        }
    }
}

func applyCatalogFilterSort(
    items: [CatalogListItem],
    filter: ShopFilterState,
    sort: ShopSortOption
) -> [CatalogListItem] {
    var out = items
    if filter.minPrice > 0 || filter.maxPrice < 500 {
        out = out.filter { item in
            guard let usd = item.usd else { return false }
            let price = NSDecimalNumber(decimal: usd).doubleValue
            return price >= Double(filter.minPrice) && price <= Double(filter.maxPrice)
        }
    }
    if let cat = filter.category?.trimmingCharacters(in: .whitespacesAndNewlines), !cat.isEmpty {
        out = out.filter { CatalogCategoryFilter.matches(filter: cat, fields: $0.category) }
    }
    switch sort {
    case .relevance:
        return out
    case .priceAsc:
        return out.sorted {
            NSDecimalNumber(decimal: $0.usd ?? 999_999).doubleValue
                < NSDecimalNumber(decimal: $1.usd ?? 999_999).doubleValue
        }
    case .priceDesc:
        return out.sorted {
            NSDecimalNumber(decimal: $0.usd ?? 0).doubleValue
                > NSDecimalNumber(decimal: $1.usd ?? 0).doubleValue
        }
    case .nameAsc:
        return out.sorted { $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending }
    }
}
